lib/src/client/http2_connection.dart

```diff
@@ -351,39 +353,67 @@ class Http2ClientConnection implements connection.ClientConnection {
   }
 }
 
-class _SocketTransportConnector implements ClientTransportConnector {
+class SocketTransportConnector implements ClientTransportConnector {
   /// Either [InternetAddress] or [String].
   final Object _host;
   final int _port;
   final ChannelOptions _options;
-  late Socket _socket; // ignore: close_sinks
+  late Socket socket;
 
-  _SocketTransportConnector(this._host, this._port, this._options)
+  Proxy get proxy => _options.proxy;
+  Object get host => proxy.isDirect ? _host : proxy.host;
+  int get port => proxy.isDirect ? _port : proxy.port;
+
+  SocketTransportConnector(this._host, this._port, this._options)
       : assert(_host is InternetAddress || _host is String);
 
   @override
   Future<ClientTransportConnection> connect() async {
     final securityContext = _options.credentials.securityContext;
-    _socket =
-        await Socket.connect(_host, _port, timeout: _options.connectTimeout);
+    socket = await initSocket(host, port);
+
+    Stream<List<int>> incoming;
```

Comment:
It seems that in both cases code just user `initSocket` + `connectToProxy`, so maybe just do this:

```dart
Future<Stream<int>> connectImpl() async {
  socket = await Socket.connect(connectionHost, connectionPort, timeout: _options.connectTimeout);
  if (proxy.isDirect) {
    return socket;
  }
  
  // We need to ask the proxy to connect to our destination host.
  return connectThroughProxy(); 
}
```